packages/core-ethereum/crates/core-ethereum-misc/src/redeem.rs

```diff
@@ -0,0 +1,413 @@
+use utils_log::{debug, error, info};
+use utils_types::primitives::Address;
+use utils_types::traits::ToHex;
+
+lazy_static::lazy_static! {
+    /// Used as a placeholder when the redeem transaction has not yet been published on-chain
+    static ref EMPTY_TX_HASH: Hash = Hash::default();
+}
+
+pub async fn redeem_all_tickets<Db, F>(
+    db: Arc<RwLock<Db>>,
+    self_addr: &Address,
+    onchain_tx_sender: &impl Fn(AcknowledgedTicket) -> F,
+) -> Result<()>
+where
+    Db: HoprCoreEthereumDbActions,
+    F: futures::Future<Output = std::result::Result<String, String>>,
+{
+    let incoming_channels = db.read().await.get_channels_to(self_addr).await?;
+    debug!(
+        "starting to redeem all tickets in {} incoming channels to us.",
+        incoming_channels.len()
+    );
+
+    for channel in incoming_channels.iter() {
+        let channel_id = channel.get_id();
+        if let Err(e) = redeem_tickets_in_channel(db.clone(), channel, onchain_tx_sender).await {
+            error!("failed to redeem tickets in channel {channel_id}: {e}");
+        }
+    }
+
+    Ok(())
+}
+
+async fn set_being_redeemed<Db>(db: &mut Db, ack_ticket: &mut AcknowledgedTicket, tx_hash: Hash) -> Result<()>
+where
+    Db: HoprCoreEthereumDbActions,
+{
+    match ack_ticket.status {
+        Untouched => {
+            let dst = db
+                .get_channels_domain_separator()
+                .await
+                .and_then(|separator| separator.ok_or(DbError::NotFound))?;
+
+            // Check if we're going to redeem a winning ticket
+            if !ack_ticket.is_winning_ticket(&dst) {
+                return Err(NotAWinningTicket);
+            }
+        }
+        BeingAggregated { .. } => return Err(WrongTicketState(ack_ticket.to_string())),
+        BeingRedeemed { .. } => {} // Just let the TX hash to be updated
+    }
+
+    ack_ticket.status = BeingRedeemed { tx_hash };
+    debug!(
+        "setting a winning {} as being redeemed with TX hash {tx_hash}",
+        ack_ticket.ticket
+    );
+    Ok(db.update_acknowledged_ticket(ack_ticket).await?)
+}
+
+pub async fn redeem_tickets_with_counterparty<Db, F>(
+    db: Arc<RwLock<Db>>,
+    counterparty: &Address,
+    onchain_tx_sender: &impl Fn(AcknowledgedTicket) -> F,
+) -> Result<()>
+where
+    Db: HoprCoreEthereumDbActions,
+    F: futures::Future<Output = std::result::Result<String, String>>,
+{
+    let ch = db.read().await.get_channel_from(counterparty).await?;
+    if let Some(channel) = ch {
+        redeem_tickets_in_channel(db, &channel, onchain_tx_sender).await
+    } else {
+        Err(InvalidArguments(format!("cannot find channel with {counterparty}")))
+    }
+}
+
+pub async fn redeem_tickets_in_channel<Db, F>(
+    db: Arc<RwLock<Db>>,
+    channel: &ChannelEntry,
+    onchain_tx_sender: &impl Fn(AcknowledgedTicket) -> F,
+) -> Result<()>
+where
+    Db: HoprCoreEthereumDbActions,
+    F: futures::Future<Output = std::result::Result<String, String>>,
```

Comment:
Isn't there a pin missing somewhere?

Reply:
Is it necessary here? The Future is pinned is later down in the WASM binding function.